File: 5_map/Dictionary.ts
/**
 * A dictionary that maps values to unique ids.
 */
class Dictionary<V> {
  private readonly _valueToId = new Map<V, number>()
  private readonly _idToValue: V[] = []

  id(value: V): number {
    const res = this._valueToId.get(value)
    if (res !== void 0) return res
    const id = this._idToValue.length
    this._idToValue.push(value)
    this._valueToId.set(value, id)
    return id
  }

  value(id: number): V | undefined {
    if (id < 0 || id >= this._idToValue.length) return void 0
    return this._idToValue[id]
  }

  get size(): number {
    return this._idToValue.length
  }
}

export { Dictionary }

if (require.main === module) {
  const dict = new Dictionary<string>()
  console.log(dict.id('a'))
  console.log(dict.id('b'))
  console.log(dict.id('a'))
  console.log(dict.value(0))
  console.log(dict.value(1))
  console.log(dict.value(2))
  console.log(dict.size)
}

File: 5_map/Dictionary.go
package main

import "fmt"

func main() {
	dict := NewDictionary()
	fmt.Println(dict.Size(), dict.Id("a"), dict.Id("b"), dict.Id("c"))
	fmt.Println(dict.Value(0))
	fmt.Println(dict.Value(10))
}

type V = interface{}
type Dictionary struct {
	_idToValue []V
	_valueToId map[V]int
}

// A dictionary that maps values to unique ids.
func NewDictionary() *Dictionary {
	return &Dictionary{
		_valueToId: map[V]int{},
	}
}

func (d *Dictionary) Id(value V) int {
	res, ok := d._valueToId[value]
	if ok {
		return res
	}
	id := len(d._idToValue)
	d._idToValue = append(d._idToValue, value)
	d._valueToId[value] = id
	return id
}

func (d *Dictionary) Value(id int) V {
	return d._idToValue[id]
}

func (d *Dictionary) Size() int {
	return len(d._idToValue)
}
